Source/DafnyLanguageServer/Handlers/DafnyCodeActionHandler.cs

```diff
@@ -0,0 +1,157 @@
+using Microsoft.Extensions.Logging;
+using OmniSharp.Extensions.LanguageServer.Protocol.Client.Capabilities;
+using OmniSharp.Extensions.LanguageServer.Protocol.Document;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Dafny.LanguageServer.Plugins;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Dafny.LanguageServer.Handlers; 
+
+public class DafnyCodeActionHandler : CodeActionHandlerBase {
+  private readonly ILogger<DafnyCodeActionHandler> logger;
+  private readonly IDocumentDatabase documents;
+
+  public DafnyCodeActionHandler(ILogger<DafnyCodeActionHandler> logger, IDocumentDatabase documents) {
+    this.logger = logger;
+    this.documents = documents;
+  }
+
+  public record QuickFixWithId(QuickFix QuickFix, int Id);
+
+  protected override CodeActionRegistrationOptions CreateRegistrationOptions(CodeActionCapability capability,
+    ClientCapabilities clientCapabilities) {
+    return new CodeActionRegistrationOptions {
+      DocumentSelector = DocumentSelector.ForLanguage("dafny"),
+      ResolveProvider = true,
+      CodeActionKinds = Container<CodeActionKind>.From(
+        CodeActionKind.QuickFix
+      ),
+      WorkDoneProgress = false
+    };
+  }
+
+
+  /// <summary>
+  /// Returns the fixes along with a unique identifier
+  /// </summary>
+  public IEnumerable<QuickFixWithId> GetFixesWithIds(QuickFixer[] fixers, DafnyDocument document, CodeActionParams request) {
+    var id = 0;
+    foreach (var fixer in fixers) {
+      var fixerInput = new VerificationQuickFixerInput(document);
+      var quickFixes = fixer.GetQuickFixes(fixerInput, request.Range);
+      var fixerCodeActions = quickFixes.Select(quickFix =>
+        new QuickFixWithId(quickFix, id++));
+      foreach (var codeAction in fixerCodeActions) {
+        yield return codeAction;
+      }
+    }
+  }
+
+  private readonly ConcurrentDictionary<string, IReadOnlyList<QuickFixWithId>> documentUriToQuickFixes = new();
+
+  public override async Task<CommandOrCodeActionContainer> Handle(CodeActionParams request, CancellationToken cancellationToken) {
+    var document = await documents.GetLastDocumentAsync(request.TextDocument);
+    if (document == null) {
+      logger.LogWarning("quick fixes requested for unloaded document {DocumentUri}", request.TextDocument.Uri);
+      return new CommandOrCodeActionContainer();
+    }
+    var quickFixers = GetQuickFixers();
+    var fixesWithId = GetFixesWithIds(quickFixers, document, request).ToArray();
+
+    var documentUri = document.Uri.ToString();
+    documentUriToQuickFixes.AddOrUpdate(documentUri,
+      _ => fixesWithId, (_, _) => fixesWithId);
+    var codeActions = fixesWithId.Select(fixWithId => {
+      CommandOrCodeAction t = new CodeAction {
+        Title = fixWithId.QuickFix.Title,
+        Data = new JArray(documentUri, fixWithId.Id)
```

Comment:
One of the fields of a CodeAction is `diagnostics`, and filling this field causes the code action to show up grouped under the diagnostic in the Problems view. The code action provided by `VerificationQuickFixer` relates to a diagnostic, so shouldn't we connect those two using the API ?